File: app/src/main/java/com/dyaco/spirit_commercial/support/room/spirit/SpiritDao.java
package com.dyaco.spirit_commercial.support.room.spirit;


import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import com.dyaco.spirit_commercial.support.room.spirit.spirit_entity.DeviceEntity;
import com.dyaco.spirit_commercial.support.room.spirit.spirit_entity.EgymEntity;
import com.dyaco.spirit_commercial.support.room.spirit.spirit_entity.ErrorMsgEntity;
import com.dyaco.spirit_commercial.support.room.spirit.spirit_entity.MediaAppsEntity;
import com.dyaco.spirit_commercial.support.room.spirit.spirit_entity.RankEntity;
import com.dyaco.spirit_commercial.support.room.spirit.spirit_entity.UploadWorkoutDataEntity;
import com.dyaco.spirit_commercial.support.room.spirit.spirit_entity.UserProfileEntity;

import java.util.List;

import io.reactivex.Maybe;

@Dao
public interface SpiritDao {
    @Query("select * from " + UserProfileEntity.USER_PROFILE + " where userType = 1")
    Maybe<List<UserProfileEntity>> getUserProfiles();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insertUserProfile(UserProfileEntity userProfileEntity);


    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insertUploadWorkoutData(UploadWorkoutDataEntity uploadWorkoutDataEntity);

    @Delete
    void deleteUploadWorkoutData(UploadWorkoutDataEntity uploadWorkoutDataEntity);

    @Transaction
    @Query("SELECT * FROM " + UploadWorkoutDataEntity.UPLOAD_WORKOUT_DATA)
    Maybe<List<UploadWorkoutDataEntity>> getUploadWorkoutDataList();


    @Update
    void updateUserProfile(UserProfileEntity userProfileEntity);

    @Query("select * from " + UserProfileEntity.USER_PROFILE + " where userType = 0")
    Maybe<List<UserProfileEntity>> getUserProfilesGuest();

    @Delete
    void deleteUserProfile(UserProfileEntity userProfileEntity);


    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insertErrorMsg(ErrorMsgEntity errorMsgEntity);

    @Transaction
    @Query("SELECT * FROM " + ErrorMsgEntity.ERROR_MSG + " ORDER BY errorDate DESC")
    Maybe<List<ErrorMsgEntity>> getErrorMsgList();

    @Query("SELECT * FROM " + ErrorMsgEntity.ERROR_MSG + " where errorCode = :errorCode ORDER BY errorDate DESC")
    Maybe<ErrorMsgEntity> getErrorMsgByErrorCode(String errorCode);


    @Delete
    void deleteErrorMsg(ErrorMsgEntity errorMsgEntity);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insertRankData(RankEntity rankEntity);

    @Transaction
    @Query("SELECT * FROM " + RankEntity.RANKS + " ORDER BY memberNo DESC")
    Maybe<List<RankEntity>> getRankList();


    @Update
    void updateDeviceDate(DeviceEntity deviceEntity);

    @Query("SELECT deviceId FROM " + DeviceEntity.DEVICE + " where deviceId = 0")
    Maybe<Integer> checkDevice();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insertDeviceDate(DeviceEntity deviceEntity);

    @Query("select * from " + DeviceEntity.DEVICE + " where deviceId = 0")
    Maybe<DeviceEntity> getDeviceData();

    @Delete
    void deleteDeviceData(DeviceEntity deviceEntity);




    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insertMediasEntity(MediaAppsEntity mediaAppsEntity);
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertMediaAppList(List<MediaAppsEntity> mediaAppsEntityList);

    @Transaction
    @Query("SELECT * FROM " + MediaAppsEntity.MEDIA_APPS + " where packageName = :packageName ORDER BY sort DESC")
    Maybe<MediaAppsEntity> getMediasAppByPackageName(String packageName);

    @Transaction
    @Query("SELECT * FROM " + MediaAppsEntity.MEDIA_APPS + " ORDER BY sort DESC")
    Maybe<List<MediaAppsEntity>> getAllMediasAppList();

    @Delete
    void deleteMediaApp(MediaAppsEntity mediaAppsEntity);

    @Query("DELETE FROM media_apps_entity")
    void deleteAllMediaApp();

    @Update
    void updateMediaApp(MediaAppsEntity mediaAppsEntity);


    @Query("UPDATE media_apps_entity set sort = :sort where packageName = :packageName")
    void updateMediaAppSort(String packageName,int sort);




    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long insertEgymData(EgymEntity egymEntity);

    @Transaction
    @Query("SELECT * FROM " + EgymEntity.EGYM_TABLE + " ORDER BY updateTime DESC")
    Maybe<List<EgymEntity>> getEgymDataList();

    @Query("SELECT * FROM " + EgymEntity.EGYM_TABLE + " where uid = :uid ORDER BY uid DESC")
    Maybe<EgymEntity> getEgymDataByUid(long uid);


    @Delete
    void deleteEgymData(EgymEntity egymEntity);
}

File: app/src/main/java/com/dyaco/spirit_commercial/support/room/spirit/SpiritDao.kt
package com.dyaco.spirit_commercial.support.room.spirit

import androidx.room.*
import com.dyaco.spirit_commercial.support.room.spirit.spirit_entity.*
import io.reactivex.Maybe

@Dao
interface SpiritDao {
    @Query("select * from " + UserProfileEntity.USER_PROFILE + " where userType = 1")
    fun getUserProfiles(): Maybe<List<UserProfileEntity>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertUserProfile(userProfileEntity: UserProfileEntity): Long

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertUploadWorkoutData(uploadWorkoutDataEntity: UploadWorkoutDataEntity): Long

    @Delete
    fun deleteUploadWorkoutData(uploadWorkoutDataEntity: UploadWorkoutDataEntity)

    @Transaction
    @Query("SELECT * FROM " + UploadWorkoutDataEntity.UPLOAD_WORKOUT_DATA)
    fun getUploadWorkoutDataList(): Maybe<List<UploadWorkoutDataEntity>>

    @Update
    fun updateUserProfile(userProfileEntity: UserProfileEntity)

    @Query("select * from " + UserProfileEntity.USER_PROFILE + " where userType = 0")
    fun getUserProfilesGuest(): Maybe<List<UserProfileEntity>>

    @Delete
    fun deleteUserProfile(userProfileEntity: UserProfileEntity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertErrorMsg(errorMsgEntity: ErrorMsgEntity): Long

    @Transaction
    @Query("SELECT * FROM " + ErrorMsgEntity.ERROR_MSG + " ORDER BY errorDate DESC")
    fun getErrorMsgList(): Maybe<List<ErrorMsgEntity>>

    @Query("SELECT * FROM " + ErrorMsgEntity.ERROR_MSG + " where errorCode = :errorCode ORDER BY errorDate DESC")
    fun getErrorMsgByErrorCode(errorCode: String): Maybe<ErrorMsgEntity>

    @Delete
    fun deleteErrorMsg(errorMsgEntity: ErrorMsgEntity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertRankData(rankEntity: RankEntity): Long

    @Transaction
    @Query("SELECT * FROM " + RankEntity.RANKS + " ORDER BY memberNo DESC")
    fun getRankList(): Maybe<List<RankEntity>>

    @Update
    fun updateDeviceDate(deviceEntity: DeviceEntity)

    @Query("SELECT deviceId FROM " + DeviceEntity.DEVICE + " where deviceId = 0")
    fun checkDevice(): Maybe<Int>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertDeviceDate(deviceEntity: DeviceEntity): Long

    @Query("select * from " + DeviceEntity.DEVICE + " where deviceId = 0")
    fun getDeviceData(): Maybe<DeviceEntity>

    @Delete
    fun deleteDeviceData(deviceEntity: DeviceEntity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertMediasEntity(mediaAppsEntity: MediaAppsEntity): Long

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertMediaAppList(mediaAppsEntityList: List<MediaAppsEntity>)

    @Transaction
    @Query("SELECT * FROM " + MediaAppsEntity.MEDIA_APPS + " where packageName = :packageName ORDER BY sort DESC")
    fun getMediasAppByPackageName(packageName: String): Maybe<MediaAppsEntity>

    @Transaction
    @Query("SELECT * FROM " + MediaAppsEntity.MEDIA_APPS + " ORDER BY sort DESC")
    fun getAllMediasAppList(): Maybe<List<MediaAppsEntity>>

    @Delete
    fun deleteMediaApp(mediaAppsEntity: MediaAppsEntity)

    @Query("DELETE FROM media_apps_entity")
    fun deleteAllMediaApp()

    @Update
    fun updateMediaApp(mediaAppsEntity: MediaAppsEntity)

    @Query("UPDATE media_apps_entity set sort = :sort where packageName = :packageName")
    fun updateMediaAppSort(packageName: String, sort: Int)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertEgymData(egymEntity: EgymEntity): Long

    @Transaction
    @Query("SELECT * FROM " + EgymEntity.EGYM_TABLE + " ORDER BY updateTime DESC")
    fun getEgymDataList(): Maybe<List<EgymEntity>>

    @Query("SELECT * FROM " + EgymEntity.EGYM_TABLE + " where uid = :uid ORDER BY uid DESC")
    fun getEgymDataByUid(uid: Long): Maybe<EgymEntity>

    @Delete
    fun deleteEgymData(egymEntity: EgymEntity)
}